File: Day3.py
import copy


def useful_data_out_for_position(bin_list, position):
    """
    Function that takes a list of binary numbers and gets some useful data out for a certain position
    :param bin_list: string list of binary numbers
    :param position: requested position of binary number to look at
    :returns:
        zeros - number of zeros in that position
        ones - number of ones in that position
        z_list - sublist of binary numbers with a zero in the requested position
        o_list - sublist of binary numbers with a one in the requested position
    """
    zeros = 0
    ones = 0
    z_list = list()
    o_list = list()
    for x in bin_list:
        x = x.strip()
        if x[position] == '0':
            zeros += 1
            z_list.append(x)
        elif x[position] == '1':
            ones += 1
            o_list.append(x)
    return zeros, ones, z_list, o_list


# Part 1
f = open("Day3.txt")
binary_list = f.readlines()
num_bits = 12

# create a dictionary to contain bit info data
bit_info = {}
for i in range(1, num_bits + 1):
    # dictionary format - the key is bit position, and the list data is [number of zeros, number of ones]
    bit_info["bit{0}".format(i)] = [0, 0]

# populate data dictionary using function
for j in range(num_bits):
    data = useful_data_out_for_position(binary_list, j)
    bit_info["bit{0}".format(j + 1)][0] = data[0]
    bit_info["bit{0}".format(j + 1)][1] = data[1]

# read the information in data dictionary to decode data
gamma_rate = ''
for k in range(1, num_bits + 1):
    if bit_info["bit{0}".format(k)][0] > bit_info["bit{0}".format(k)][1]:
        gamma_rate += '0'
    else:
        gamma_rate += '1'
gamma_rate_dec = int(gamma_rate, 2)
print(f"The gamma rate: binary = {gamma_rate}, decimal = {gamma_rate_dec}")

# flip bits from gamma rate for epsilon rate and convert to decimal
epsilon_rate = ''.join(['1' if i == '0' else '0' for i in gamma_rate])
epsilon_rate_dec = int(epsilon_rate, 2)
print(f"The epsilon rate: binary = {epsilon_rate}, decimal = {epsilon_rate_dec}")

# multiply two rates in binary and get binary number output
power_consumption = bin(gamma_rate_dec * epsilon_rate_dec)
power_consumption_dec = gamma_rate_dec * epsilon_rate_dec
print(f"The power consumption: binary = {power_consumption}, decimal = {power_consumption_dec}\n")


# Part 2
# use the function to loop over a filtered list until the filtered list becomes one value for oxygen generator rating
filtered_list = copy.deepcopy(binary_list)
m = 0
while len(filtered_list) != 1:
    num_zeros, num_ones, zeros_list, ones_list = useful_data_out_for_position(filtered_list, m)
    if num_zeros > num_ones:
        filtered_list = zeros_list
        m += 1
    elif num_ones > num_zeros:
        filtered_list = ones_list
        m += 1
    elif num_ones == num_zeros:
        filtered_list = ones_list
        m += 1
oxygen_generator_rating = filtered_list[0]  # filtered_list should only contain one value now
oxygen_generator_rating_dec = int(oxygen_generator_rating, 2)
print(f"The oxygen generator rating: binary = {oxygen_generator_rating}, decimal = {oxygen_generator_rating_dec}")

# use the function to loop over a filtered list until the filtered list becomes one value for c02 scrubber rating
# essentially same code as above but switch the list that gets assigned to filtered list
filtered_list = copy.deepcopy(binary_list)
m = 0
while len(filtered_list) != 1:
    num_zeros, num_ones, zeros_list, ones_list = useful_data_out_for_position(filtered_list, m)
    if num_zeros > num_ones:
        filtered_list = ones_list  # switch what list we assign for c02 scrubber rating
        m += 1
    elif num_ones > num_zeros:
        filtered_list = zeros_list  # switch what list we assign for c02 scrubber rating
        m += 1
    elif num_ones == num_zeros:
        filtered_list = zeros_list  # switch what list we assign for c02 scrubber rating
        m += 1
c02_scrubber_rating = filtered_list[0]  # filtered_list should only contain one value now
c02_scrubber_rating_dec = int(c02_scrubber_rating, 2)
print(f"The c02 scrubber rating: binary = {c02_scrubber_rating}, decimal = {c02_scrubber_rating_dec}")

life_support_rating = bin(oxygen_generator_rating_dec * c02_scrubber_rating_dec)
life_support_rating_dec = oxygen_generator_rating_dec * c02_scrubber_rating_dec
print(f"The life support rating: binary = {life_support_rating}, decimal = {life_support_rating_dec}")

File: Port to C++/Day3.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <numeric>
#include <bitset>

using namespace std;

auto useful_data_out_for_position(vector<string> bin_list, int position) {
    int zeros = 0;
    int ones = 0;
    vector<string> z_list;
    vector<string> o_list;
    for (auto x : bin_list) {
        //x = x.strip()
        if (x[position] == '0') {
            zeros += 1;
            z_list.push_back(x);
        }
        else if (x[position] == '1') {
            ones += 1;
            o_list.push_back(x);
        }            
    }
    
    struct ReturnVals {
        int _1;
        int _2;
        vector<string> _3;
        vector<string> _4;
    };
    return ReturnVals {zeros, ones, z_list, o_list};
}

int main() {
    //Get the txt input
    ifstream myfile("..\\Personal txt Inputs\\Day3.txt");   //create an input stream object from file
    string file_line;                                            //create line object to read into
    vector<string> binary_list;                            //create vector of ints/strings
    while(getline(myfile, file_line)) {                          //read each line and cast to file_line variable
        binary_list.push_back(file_line);                       //append file_line variable to vector
    }
    myfile.close();

    int num_bits = 12;

    vector<int> bit_info[num_bits];  // create array of vector strings as python dict equivalent, index is bit num, vector contains info
    // skip counter i used in python version as we do not bother initialising dict values to [0,0] like we do there

    for (int j = 0; j < num_bits; ++j) {
        auto [num_zeros, num_ones, zeros_list, ones_list] = useful_data_out_for_position(binary_list, j);
        bit_info[j].push_back(num_zeros);
        bit_info[j].push_back(num_ones);
    }

    string gamma_rate;
    for (int k = 0; k < num_bits; ++k) {

        if (bit_info[k][0] > bit_info[k][1]) {
            gamma_rate += '0';
        }
        else {
            gamma_rate += '1';
        }
    }
    int gamma_rate_dec = stoi(gamma_rate, nullptr, 2);
    cout << "The gamma rate: binary = " << gamma_rate << ", decimal = " << gamma_rate_dec << endl;

    string epsilon_rate;
    for (char i : gamma_rate) {
        if (i == '0'){
            epsilon_rate += '1';
        } else {
            epsilon_rate += '0';
        }
    }
    int epsilon_rate_dec = stoi(epsilon_rate, nullptr, 2);
    cout << "The epsilon_rate: binary = " << epsilon_rate << ", decimal = " << epsilon_rate_dec << endl;
    
    int power_consumption_dec = gamma_rate_dec * epsilon_rate_dec;
    string power_consumption = bitset<24>(gamma_rate_dec * epsilon_rate_dec).to_string();
    cout << "The power_consumption: binary = " << power_consumption << ", decimal = " << power_consumption_dec << endl << endl;


    // Part 2 
    // use the function to loop over a filtered list until the filtered list becomes one value for oxygen generator rating
    vector<string> filtered_list = binary_list;  
    int m = 0;
    while (filtered_list.size() != 1) {
        auto [num_zeros, num_ones, zeros_list, ones_list] = useful_data_out_for_position(filtered_list, m);
        if (num_zeros > num_ones) {
            filtered_list = zeros_list;
            m += 1;
        }
        else if (num_ones > num_zeros) {
            filtered_list = ones_list;
            m += 1;
        }
        else if (num_ones == num_zeros) {
            filtered_list = ones_list;
            m += 1;
        }
    }

    string oxygen_generator_rating = filtered_list[0];  // filtered_list should only contain one value now
    int oxygen_generator_rating_dec = stoi(oxygen_generator_rating, nullptr, 2);
    cout << "The oxygen_generator_rating: binary = " << oxygen_generator_rating << ", decimal = " << oxygen_generator_rating_dec << endl;

    // use the function to loop over a filtered list until the filtered list becomes one value for c02 scrubber rating
    // essentially same code as above but switch the list that gets assigned to filtered list
    filtered_list = binary_list;  
    m = 0;
    while (filtered_list.size() != 1) {
        auto [num_zeros, num_ones, zeros_list, ones_list] = useful_data_out_for_position(filtered_list, m);
        if (num_zeros > num_ones) {
            filtered_list = ones_list;  // switch what list we assign for c02 scrubber rating
            m += 1;
        }
        else if (num_ones > num_zeros) {
            filtered_list = zeros_list;  // switch what list we assign for c02 scrubber rating
            m += 1;
        }
        else if (num_ones == num_zeros) {
            filtered_list = zeros_list;  // switch what list we assign for c02 scrubber rating
            m += 1;
        }
    }

    string c02_scrubber_rating = filtered_list[0];  // filtered_list should only contain one value now
    int c02_scrubber_rating_dec = stoi(c02_scrubber_rating, nullptr, 2);
    cout << "The c02_scrubber_rating: binary = " << c02_scrubber_rating << ", decimal = " << c02_scrubber_rating_dec << endl;

    int life_support_rating_dec = oxygen_generator_rating_dec * c02_scrubber_rating_dec;
    string life_support_rating = bitset<25>(oxygen_generator_rating_dec * c02_scrubber_rating_dec).to_string();
    cout << "The life_support_rating: binary = " << life_support_rating << ", decimal = " << life_support_rating_dec << endl;

};

